crates/bevy_pbr/src/material.rs

```diff
@@ -937,34 +975,107 @@ pub fn check_entities_needing_specialization<M>(
+    render_materials: Res<RenderAssets<PreparedMaterial<M>>>,
+    render_mesh_instances: Res<RenderMeshInstances>,
+    render_meshes: Res<RenderAssets<RenderMesh>>,
+    material_bind_group_allocator: Res<MaterialBindGroupAllocator<M>>,
+    render_lightmaps: Res<RenderLightmaps>,
+    render_visibility_ranges: Res<RenderVisibilityRanges>,
+    ticks: SystemChangeTick,
+) where
+    M::Data: PartialEq + Eq + Hash + Clone,
+{
+    for entity in &entities_maybe_needing_specialization.entities {
+        let Some(render_material_instance) = render_material_instances.instances.get(entity) else {
+            continue;
+        };
+        let Ok(material_asset_id) = render_material_instance.asset_id.try_typed::<M>() else {
+            continue;
+        };
+        let Some(material) = render_materials.get(material_asset_id) else {
+            continue;
+        };
+        let Some(mesh_instance) = render_mesh_instances.render_mesh_queue_data(*entity) else {
+            continue;
+        };
+        let Some(mesh) = render_meshes.get(mesh_instance.mesh_asset_id) else {
+            continue;
+        };
+        let Some(material_bind_group) = material_bind_group_allocator.get(material.binding.group)
+        else {
+            continue;
+        };
+
+        let mesh_pipeline_key_bits = material.properties.mesh_pipeline_key_bits;
+        let mut mesh_key =
+            MeshPipelineKey::from_bits_retain(mesh.key_bits.bits()) | mesh_pipeline_key_bits;
+        if let Some(lightmap) = render_lightmaps.render_lightmaps.get(entity) {
+            mesh_key |= MeshPipelineKey::LIGHTMAPPED;
+
+            if lightmap.bicubic_sampling {
+                mesh_key |= MeshPipelineKey::LIGHTMAP_BICUBIC_SAMPLING;
+            }
+        }
+        if render_visibility_ranges.entity_has_crossfading_visibility_ranges(*entity) {
+            mesh_key |= MeshPipelineKey::VISIBILITY_RANGE_DITHER;
+        }
+
+        let material_key = material_bind_group
+            .get_extra_data(material.binding.slot)
+            .clone();
+        if let Some(prev_material_key) = material_key_cache.entities.get(entity) {
+            if *prev_material_key != material_key {
+                material_key_cache.insert(*entity, material_key);
+                entity_specialization_ticks.insert(*entity, ticks.this_run());
+            }
+        } else {
+            material_key_cache.entities.insert(*entity, material_key);
+            entity_specialization_ticks.insert(*entity, ticks.this_run());
+        }
+        if let Some(prev_mesh_key) = mesh_key_cache.get(entity) {
+            if *prev_mesh_key != mesh_key {
+                mesh_key_cache.insert(*entity, mesh_key);
+                entity_specialization_ticks.insert(*entity, ticks.this_run());
+            }
+        } else {
+            mesh_key_cache.insert(*entity, mesh_key);
+            entity_specialization_ticks.insert(*entity, ticks.this_run());
```

Comment:
Isn’t the point to skip specialisation of the key didn’t change? One or the other of these branches looks wrong. Either this change tick was incremented suggesting that the key changed this run so specialisation is needed, which is wrong? Or the other branch incremented the change tick to indicate specialisation isn’t needed, which is wrong? Or I’m wrong :)